tests/test_ote_training.py

```diff
@@ -0,0 +1,613 @@
+        logger.info(f'performance={self.output_model.performance}')
+
+    def get_training_performance(self):
+        return getattr(self.output_model, 'performance', None)
+
+    def run_ote_training_once(self, data_collector):
+        if self.was_training_run and self.stored_exception_training:
+            logger.warn('In function run_ote_training_once: found that previous call of the function '
+                        'caused exception -- re-raising it')
+            raise self.stored_exception_training
+
+        if not self.was_training_run:
+            try:
+                self._run_ote_training()
+                self.was_training_run = True
+            except Exception as e:
+                self.stored_exception_training = e
+                self.was_training_run = True
+                raise e
+
+        training_performance = self.get_training_performance()
+
+        score_name, score_value = performance_to_score_name_value(training_performance)
+        if score_name:
+            data_collector.log_final_metric('training_performance/' + score_name, score_value)
+        else:
+            logger.warning(f'WARNING: Cannot get training performance')
+
+        hyperparams_dict = convert_hyperparams_to_dict(self.copy_hyperparams)
+        for k, v in hyperparams_dict.items():
+            data_collector.update_metadata(k, v)
+
+        return training_performance
+
+    def run_ote_evaluation(self, data_collector, subset=Subset.VALIDATION):
+        if not self.was_training_run:
+            raise RuntimeError('Training was not run for the OTETrainingImpl instance')
+        if self.stored_exception_training:
+            raise RuntimeError('Training was not successful for the OTETrainingImpl instance')
+        logger.debug('Get predictions on the validation set')
+        validation_dataset = self.dataset.get_subset(subset)
+        self.predicted_validation_dataset = self.task.infer(
+            validation_dataset.with_empty_annotations(),
+            InferenceParameters(is_evaluation=True))
+        self.resultset = ResultSetEntity(
+            model=self.output_model,
+            ground_truth_dataset=validation_dataset,
+            prediction_dataset=self.predicted_validation_dataset,
+        )
+        logger.debug('Estimate quality on validation set')
+        self.task.evaluate(self.resultset)
+        self.evaluation_performance = self.resultset.performance
+        logger.info(f'performance={self.evaluation_performance}')
+        score_name, score_value = performance_to_score_name_value(self.evaluation_performance)
+        data_collector.log_final_metric('evaluation_performance/' + score_name, score_value)
+        return self.evaluation_performance
+
+    def _run_ote_export(self, data_collector):
+        logger.debug('Copy environment for evaluation exported model')
+        # TODO(lbeynens): CONSIDER IT WITH Pavel
+        self.environment_for_export = copy.copy(self.environment)
```

Comment:
Could you please elaborate, what is supposed to be considered here?

Reply:
The point that I'd like to consider is: if we should do `copy.copy` here? or`copy.deepcopy`? or just `self.environment` without copy?
The reason why I'd like to consider this is as follows:
* Earlier some action with `OTEDetectionTask` caused changes in the environment's field.
  Even now the method `OTEDetectionTask.evaluate` contains the commented lines
  ```
              # only set configurable params based on validation result set
              <...>
                    # params.postprocessing.confidence_threshold = best_confidence_threshold
              <...>
            # self._task_environment.set_configurable_parameters(params)
    ```
* I do not want to have side effects that may happen if OpenVINO export is used 
  (At the moment these side effects do not occur; but I do not want to play guessing games.)
* So, I decided to make a shallow copy of the environment

Today I re-investigated how `OTEDetectionTask`  uses the environment now and found out that 
1. At the moment `OTEDetectionTask` uses environment in "read-only mode".
  In this case I could skip copying of the environment.
2. But the commented lines stated above give me a sign that it can be changed in the future
  I re-investigated and now I think that deep copy is the best choice in this case.

So, Pavel, my question to you (as the owner of `OTEDetectionTask`) is: are you going to use environment in the class in read-only mode? or you can change the code (and make it change the environment)?
What could you recommend?

Reply:
1. The change has already happened a while ago. The convention is that environment is immutable for the task. Though, there are no technical constraints to forbid this. So, it's better not rely on this fact.
1. I don't know if shallow or deep copy should be used.

Reply:
Ok, thank you.
I will use a "no copy" approach and will write a comment on possible side effects if some operation changes the environment.

Reply:
Done